### ISSUE TITLE
Rating system

### ISSUE THREAD

**Issue by mowgli** [BestImageViewer/geeqie#323]
I'd like to have some rating system for images. A lot of similar software has a five star rating system. The mark system is similar, but instead of 6 bool values, I'd like one int/float value too. Then the possibility to filter according to rating (rating < 1.5 || rating > 3).

Reported by: roggan87

**Comment by caclark** [comment]
A possibility is:
To have an additional pane on the Info sidebar called Rating, linked to Xmp.xmp.rating.
An additional Search option of "Image rating is  less than/greater than/between" etc.
An additional item on the "Sort by" button for rating.
The time to implement this is minimal - I've already done it.

The problems are:
The input box is free-text - anything can be input, but Adobe defines xmp.rating to be between -1 and 5. So any text string input as a rating would be interpreted as 0 by search/sort. (I guess filtering the input could be done with a bit of effort).
Perhaps people using this facility would not be happy having to type a number into the input box, and would only want to set Ratings using keystrokes e.g <ctrl>1 to <ctrl>5, or mouse-clicks. (Maybe the keystrokes could be fixed with a bit of effort).

A basic implementation has already been done. I'm not sure if that is usable as is, or if some other approach is necessary.

**Comment by tomaszg7** [comment]
Seems like a good idea, I recently started using similar thing for my music player (and I was looking at commits related to this the moment I got the e-mail notification about your post :) ). I agree that keyboard mapping might be essential (maybe some easy way to switch between marks and ratings, or just unify the whole thing somehow). 

An idea: I guess it is compatible with Darktable ratings, so maybe it should be good to make marks compatible with color labels in Darktable?

Do you have your patch somewhere I could try it out?

**Comment by caclark** [comment]
There is a patch here:
https://github.com/BestImageViewer/geeqie/commit/88e58a01d5cebaa5eb41d0ef00b75065fcab2c7e

The most important point is if I got the basics right - using Xmp.xmp.Rating.
If that is OK and can inter-operate with other programs, then maybe it should go into the repo.
It does not matter if, for most people, it is too difficult to use. So long as the data of anyone using it does not get lost.
It is not possible to know when, or if, I will get enthusiastic enough to improve things.

Maybe the existing coding of the marks feature could be used - just somehow switch between "map marks to keywords" and "map marks to rating".

**Comment by mowgli** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Hi,

Am Fr den  2. Jun 2017 um 12:12 schrieb Colin Clark:
> There is a patch here:
> https://github.com/BestImageViewer/geeqie/commit/88e58a01d5cebaa5eb41d0ef00b75065fcab2c7e
> 
> The most important point is if I got the basics right - using Xmp.xmp.Rating.

I am not sure if it could fit. Have to check in a quite moment.

Regards
   Klaus
- -- 
Klaus Ethgen                                       http://www.ethgen.ch/
pub  4096R/4E20AF1C 2011-05-16            Klaus Ethgen <Klaus@Ethgen.ch>
Fingerprint: 85D4 CA42 952C 949B 1753  62B3 79D0 B06F 4E20 AF1C
-----BEGIN PGP SIGNATURE-----
Comment: Charset: ISO-8859-1

iQGzBAEBCgAdFiEEMWF28vh4/UMJJLQEpnwKsYAZ9qwFAlkxTcYACgkQpnwKsYAZ
9qzghgv/Xqfub1f5j0epC4Nkyh6I6jJlgbY6/PB7bnLkiMNsz6rxcgE1ufXeFx6Z
nFCvXtr/hix1zBInpoFHdUK17AUVW1DQ6HWrA/1M3G/XMz9IT6th69X0lb7eoUnw
u06YxyE2Yj2LvC9KdNQ1oCTAFNcLUM1Wo0ufC3/rWO1IBQ+nHKfxB8KWT9z9oiTZ
d/7ObYG2bSm0mnOvO4dgbr68K7G2K8zCg7FMRcgyNRJldUKedeIp8k3ef971ezLh
RffhRjo9cWRcBnZBc0fRtWqsLdHReflE5I2QuA8Dw3H2IUm1uc+IRnbq8yXebJea
fvnCy5KXov4FGy/j8ubHC5ecdU1UBfHbQ+BkakWHyBxERZLzAywB8h7vdNByGPCR
uOkEJlFrBjZJ2DPQqdsP5+veCtfsVU401NuNUp33lvPc/S3H1lPLlgzBBU/aJpPV
B+7r+IVaN7RYde+C/0pdO1M0hPcNSiQArQOOynPtmMQtHM6B3r2Xid+6GAPeiwl8
zx9ugvvx
=m2Ln
-----END PGP SIGNATURE-----

**Comment by tomaszg7** [comment]
> There is a patch here:

Can't get it to work - the pane just doesn't show. Its height is set to 1 and if I change it to something else, upon restart of Geeqie it reverts to 1. 

BTW. Maybe it should be noted in preferences dialog that changing info sidebar heights requires restart.

> Maybe the existing coding of the marks feature could be used - just somehow switch between "map marks to keywords" and "map marks to rating".

Also it would be nice to have a possibility to disable these maps altogether without deleting them. For example, when one has set up some keywords maps, but wants to use plain marks for a while.

**Comment by caclark** [comment]
> the pane just doesn't show

Could you check what is written into geeqierc.xml for this parameter?

> Maybe it should be noted in preferences dialog that changing info sidebar heights requires restart

OK, on next commit

>  have a possibility to disable these maps altogether without deleting them

Definitely a good idea when you have connected keywords/marks, and need to temporarily disable them

**Comment by tomaszg7** [comment]
> Could you check what is written into geeqierc.xml for this parameter?

Nope. But when I add it manually, it does show and it interoperates with Darktable correctly. Not sure where the value "1" comes from, as the code uses default "10". Changing it to some bigger value like "50" didn't change that.

**Comment by caclark** [comment]
> Also it would be nice to have a possibility to disable these maps altogether without deleting them. For example, when one has set up some keywords maps, but wants to use plain marks for a while.

Refer to #489

**Comment by caclark** [comment]
> Can't get it to work - the pane just doesn't show. Its height is set to 1 and if I change it to something else, upon restart of Geeqie it reverts to 1.

The only way I can replicate this fault is to delete geeqierc.xml, start Geeqie and then close it without opening the info sidebar. Because there is no sidebar data in geeqierc.xml, the heights do not get stored.

Even if this is not the problem you have experienced, maybe that part of Preferences should be greyed out until the sidebar is opened.

**Comment by tomaszg7** [comment]
When I start Geeqie without any geeqierc.xml file, everything works as expected. But as soon as I manually delete the relevant line from this file, the problem comes back.  Just checked that the same is true with other sections like "title".

BTW. I don't like the idea of storing XMP field name as a "key" in geeqierc.xml file. Moreover it is done in a way not accessible via "preferences" dialog. I believe that either this value should be kept separate in xml file as a hidden setting, or just hardcoded.

**Comment by caclark** [comment]
> When I start Geeqie without any geeqierc.xml file, everything works as expected. But as soon as I manually delete the relevant line from this file, the problem comes back.

Unfortunately this is true. You will not be able to store a height value until you re-enable the Rating or Title pane by right-clicking and adding it to the info sidebar again.
I implemented the height values in Preferences with the intention that it was just a quick stop-gap - better than editing geeqierc.xml, but not by much. I don't think it worth expending any more effort on it, because the correct way would be to use GtkPaned widgets so that the user can simply drag the divider to set the height. One day I'll get around to it.

> BTW. I don't like the idea of storing XMP field name as a "key" in geeqierc.xml file

Yes, it doesn't make much sense for Title, Keywords and Comment to have the xmp tag in the config file - if the tags are fixed in the Xmp spec, then they should be fixed in the code and not changeable by the user.

The pane_exif entries (Exif, File info, Location and Copyright) are all pretty much free-format - for example, you can store xmp values in File info if you want. This data has to go into the config file.

[It should not be too difficult to permit the user to set up their own pane, with their own title and contents. But at the moment they can achieve the same effect by adding to the content of the Exif pane.]

I think the documentation of which exif and xmp tags are used is a bit deficient. The information is buried in the code, but it should be made clearer to the user e.g. which xmp tag is used for the Title pane etc. Maybe I can simply search the code and collect all the data into an appendix to the Help file.

**Comment by tomaszg7** [comment]
> Unfortunately this is true. You will not be able to store a height value until you re-enable the Rating or Title pane by right-clicking and adding it to the info sidebar again.

Ah, I forgot about right-clicking. That was not obvious for me, especially since you have to click on the title of existing pane, not on the empty space. Sorry. 

However it would probably be much more intuitive for users if right-clicking empty space would show "add pane" popup. It might be even better to put some kind of little "+" icon in the bottom corner of empty space.

> I implemented the height values in Preferences with the intention that it was just a quick stop-gap - better than editing geeqierc.xml, but not by much. 

Ok, I understand that. So, maybe just disable the option if there is no corresponding pane?

**Comment by caclark** [comment]
> However it would probably be much more intuitive for users if right-clicking empty space would show "add pane" popup. It might be even better to put some kind of little "+" icon in the bottom corner of empty space.

Refer to #490

**Comment by caclark** [comment]
Initial implementation in commit cb2fa6ae0 at geeqie.org.

Set values either by Edit menu, or Alt+Keypad+n: n is 0 to 5
Alt+keypad+minus sets the value to -1.

If you select sort-by-rating on a folder with a large number of files, Geeqie will freeze for a while due to the time it takes to read the XMP data (as for sort by exif date).
This stuff really needs to go into the background.

**Comment by caclark** [comment]
Commit 257cba812e at geeqie.org relates to this.

**Comment by caclark** [comment]
Attached is a .diff that shows star rating in the files pane, enabled via a right-click menu item.
The problem is that if you enable this on a folder with a few thousand images, Geeqie freezes until the exif data is read for all images.
It may be best not to include this in the repo until that problem gets solved.
[show_star_rating_1.diff.gz](https://github.com/BestImageViewer/geeqie/files/2091521/show_star_rating_1.diff.gz)

**Comment by caclark** [comment]
Commit 8be05f7705 at geeqie.org shows the star rating in the files pane.

**Comment by tomaszg7** [comment]
1. After assigning tag, you need to exit and re-enter the directory to see the rating.
2. Symbol used for showing ratings seems very exotic - I can't find a single font that can display it.

And for future - it would be good if it was possible to set ratings by clicking on the marks.

**Comment by caclark** [comment]
Commit 8be05f7705 at geeqie.org shows the star rating in the files pane.

**Comment by caclark** [comment]
> After assigning tag, you need to exit and re-enter the directory to see the rating.

Pressing Reset will do it. In icon view, just moving the mouse away from the files pane causes a refresh. It should not be difficult to fix.

> Symbol used for showing ratings seems very exotic - I can't find a single font that can display it.

Lines 304 and 311 in misc.c are where it's at. Just changing them to "-" and "*" will produce a usable response.
I'll have to find out how this Unicode/UTF8 stuff works. it was easier in the days of 8-bit ASCII.

**Comment by tomaszg7** [comment]
> Pressing Reset will do it. In icon view, just moving the mouse away from the files pane causes a refresh. It should not be difficult to fix.
Pressing refresh does the trick, but moving the mouse away does not - at least not in GTK+3 build.

> Lines 304 and 311 in misc.c are where it's at. Just changing them to "-" and "*" will produce a usable response.

I know, I checked there and my code editor was unable to display it either :) I don't even know either how UTF8 works with Glib, but I noticed `g_string_append_unichar()`. Don't know if it is better to store the "mark" character as `gunichar` or not.

**Comment by caclark** [comment]
Commit 5ccd1c0 at geeqie.org is another attempt.

Thanks for the hint.

**Comment by tomaszg7** [comment]
In any case, still doesn't show for me. From what I can see it is a glyph not present in common fonts:
https://vazor.com/unicode/c1F7CA.html
You took the one from the very end of the Unicode specification :) Maybe something from this section?
https://vazor.com/unicode/b2700.html

**Comment by caclark** [comment]
Commit 6d61e1cb6. One more time...

**Comment by tomaszg7** [comment]
Now I see it, but that's probably not a guarantee that everyone will. I don't know if there is a reasonable solution to that (apart from using plain * that is ;) ). Or a hidden configuration option?

Anyway, I noticed that "show star rating" setting is not remembered after Geeqie restart (maybe it is by design).

**Comment by caclark** [comment]
> Or a hidden configuration option?

Yes, I will make it a configurable item. Then I'll be able to use something like "Heavy Black Heart" for most on my photos.

> Anyway, I noticed that "show star rating" setting is not remembered after Geeqie restart

I coded it before I did the idle loop reads, so I thought it would be a problem to include it as a saved item.   Just one more thing to get around to. (Actually there is still a problem with program freezes when Show Star Rating is selected...)
